### PR TITLE
driver: ykushpower: support remote control via ssh.

### PR DESCRIPTION
**Pull request by chen-png**
enable driver for resource YKUSHPowerPort,
change to use ykushcmd as control driver,
support remote control via ssh.

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

**Description**

support to control YKUSHPowerPort device remotely via ssh.

### CODE REVIEW

**Comment by chen-png** [comment]
hi @Emantor @Bastian-Krause  Could you help take a look of this? thanks.
this is the first time I contacted with labgrid, but I found it solved many problems for hardware's' management and remote debug, so now I am working on using labgrid to set up a remote testing system on varieties of boards based on Zephyr OS, and found it supported yepkit switchable usb hub but didn't enable it, so I submitted this, any comments are welcome.

File: labgrid/remote/client.py [hunk 1]
```diff
@@ -763,6 +764,11 @@ def power(self):
                         drv = target.get_driver(TasmotaPowerDriver)
                     except NoDriverFoundError:
                         drv = TasmotaPowerDriver(target, name=None)
+                elif isinstance(resource, YKUSHPowerPort):
+                    try:
+                        drv = target.get_driver(YKUSHPowerDriver)
+                    except NoDriverFoundError:
+                        drv = YKUSHPowerDriver(target, name=None)
```
Comment (Bastian-Krause) [inline labgrid/remote/client.py:771]: This should be a separate commit. Also `YKUSHPowerDriver` needs to be imported first.

File: labgrid/driver/powerdriver.py [hunk 2]
```diff
@@ -239,22 +239,31 @@ class YKUSHPowerDriver(Driver, PowerResetMixin, PowerProtocol):
     delay = attr.ib(default=2.0, validator=attr.validators.instance_of(float))
 
 
+    # use ykushcmd as control driver, because this is more convenient
+    # than a python module and supports remote control via ssh.
+    # ref: https://www.yepkit.com/learn/setup-guide-ykush
```
Comment (Bastian-Krause) [inline labgrid/driver/powerdriver.py:244]: This part does not belong here, but rather in the `CHANGES.rst` file.

File: labgrid/driver/powerdriver.py [hunk 3]
```diff
@@ -239,22 +239,31 @@ class YKUSHPowerDriver(Driver, PowerResetMixin, PowerProtocol):
...
     def __attrs_post_init__(self):
         super().__attrs_post_init__()
```
Comment (Bastian-Krause) [inline labgrid/driver/powerdriver.py:246]:
```suggestion
    def __attrs_post_init__(self):
        super().__attrs_post_init__()
        if self.target.env:
            self.tool = self.target.env.config.get_tool('ykushcmd') or 'ykushcmd'
        else:
            self.tool = 'ykushcmd'

```
Let's make the CLI tool configurable. Now please use `self.tool` instead of `ykushcmd` everywhere.

File: labgrid/remote/exporter.py [hunk 4]
```diff
@@ -669,6 +669,26 @@ def _get_params(self):
...
+class YKUSHPowerPortExport(ResourceExport):
+    """ResourceExport for YKUSHPowerPort devices"""
+
+    def __attrs_post_init__(self):
+        super().__attrs_post_init__()
+        local_cls_name = self.cls
+        self.data['cls'] = f"{local_cls_name}"
+        from ..resource import ykushpowerport
+        local_cls = getattr(ykushpowerport, local_cls_name)
+        self.local = local_cls(target=None, name=None, host=self.host, **self.local_params)
```
Comment (Bastian-Krause) [inline labgrid/remote/exporter.py:682]: I don't know if that's needed or whether it's correct or not. @Emantor, @jluebbe: could you comment on this?

File: labgrid/resource/ykushpowerport.py [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
 
 from ..factory import target_factory
-from .common import Resource
+from .common import NetworkResource
 
 
 @target_factory.reg_resource
 @attr.s(eq=False)
-class YKUSHPowerPort(Resource):
+class YKUSHPowerPort(NetworkResource):
```
Comment (Bastian-Krause) [inline labgrid/resource/ykushpowerport.py:9]: Why are you changing this to a network resource rather than adding a `NetworkYKUSHPowerPort`?

File: labgrid/driver/powerdriver.py [hunk 6]
```diff
@@ -265,7 +274,24 @@ def cycle(self):
...
+        # the example of ykushcmd -g like below:
+        # cmd: ykushcmd -g 1
+        # output: Downstream port 1 is ON/OFF
+        check_str = f"port {self.port.index}"
+        if check_str in res:
+            status = res.strip().split()[-1]
+            if status == "ON":
+                return True
+            else:
+                return False
```
Comment (Bastian-Krause) [inline labgrid/driver/powerdriver.py:294]:
```suggestion
        check_str = "Downstream port {port} is {status}"
        if check_str.format(port=self.port.index, status="ON") in res:
            return True
        if check_str.format(port=self.port.index, status="OFF") in res:
            return False
            
        raise ExecutionError(f"Could not find port string in ykushcmd output: {res}")
```
This makes sure we get a return value or an exception in all cases.

**Review by Bastian-Krause — CHANGES_REQUESTED** [review]
Thanks for your contribution! See comments below.

**Comment by chen-png** [comment]
>  See comments below.

thanks for review, I will address them soon

**Comment by davidmaze1990** [comment]
Just be curious @chen-png ， have you finally setup zephyr farm with labgrid? I didn't see labgrid support pyocd or similar (correct me if I'm wrong), is the "flash zephyr image" is something you customized? Thanks!

**Comment by davidmaze1990** [comment]
And @Bastian-Krause , does labgrid support MCU world well?

**Comment by Bastian-Krause** [comment]
@davidmaze1990 Let's discuss this in #1140.

**Comment by edersondisouza** [comment]
As this PR is fairly stale, I've decided to pick this work up and created https://github.com/labgrid-project/labgrid/pull/1143.

**Comment by Emantor** [comment]
Superseded by #1143, closing.
